ui/components-react/src/test/propertygrid/MergingPropertyDataProvider.test.ts

```diff
@@ -0,0 +1,291 @@
+    const mergingProviderValue3 =
+      mergingProviderRecord3.value as PrimitiveValue;
+    const dataProvider1Value = dataProvider1Record.value as PrimitiveValue;
+    const dataProvider2Value = dataProvider2Record.value as PrimitiveValue;
+    const dataProvider3Value = dataProvider3Record.value as PrimitiveValue;
+    expect(mergingProviderValue1).to.equal(dataProvider1Value);
+    expect(mergingProviderValue2).to.equal(dataProvider2Value);
+    expect(mergingProviderValue3).to.equal(dataProvider3Value);
+  });
+
+  it("getData should return same PropertyData when no data changes", async () => {
+    const mergingProvider = createMergedPropertyDataProvider([
+      dataProvider1,
+      dataProvider2,
+      dataProvider3,
+    ]);
+    const first = await mergingProvider.getData();
+    const second = await mergingProvider.getData();
+    expect(first).to.deep.equal(second);
+  });
+
+  it("getData returns updated data when it changes", async () => {
+    const mergingProvider = createMergedPropertyDataProvider([
+      dataProvider1,
+      dataProvider2,
+      dataProvider3,
+    ]);
+    const initial = await mergingProvider.getData();
+    expect(initial.categories.length).to.equal(3);
+    dataProvider1.addCategory({
+      name: "Name1",
+      label: "Label1",
+      expand: false,
+    });
+
+    const categoryAdded = await mergingProvider.getData();
+    expect(categoryAdded.categories.length).to.equal(4);
+    // MergingProvider has 3 providers and 2nd and 3rd provider have 1 category, and 1st provider has 2 categories.
+    // MergingProvider joins all categories into a single array, since 1st provider's categories are added first, newly created
+    // category will be in the position 1.
+    expect(categoryAdded.categories[1].name).to.equal("Name1");
+
+    expect(categoryAdded.records[provider2Category.name].length).to.equal(1);
+    const property = TestUtils.createPrimitiveStringProperty(
+      "String1",
+      "string1"
+    );
+    dataProvider2.addProperty(property, 0);
+    const propertyAdded = await mergingProvider.getData();
+
+    expect(propertyAdded.records[provider2Category.name].length).to.equal(2);
+    expect(
+      (propertyAdded.records[provider2Category.name][1].value as PrimitiveValue)
+        .displayValue
+    ).to.equal("string1");
+
+    expect(propertyAdded.records[provider3Category.name].length).to.equal(1);
+    dataProvider3.removeProperty(provider3Property, 0);
+    const propertyRemoved = await mergingProvider.getData();
+    expect(propertyRemoved.records[provider3Category.name].length).to.equal(0);
+  });
+
+  it("onDataChanged should be called when any of the source providers call onDataChanged", async () => {
+    const mergingProvider = createMergedPropertyDataProvider([
+      dataProvider1,
+      dataProvider2,
+      dataProvider3,
+    ]);
+    let onDataChangedCalls = 0;
+
+    mergingProvider.onDataChanged.addListener(() => {
+      ++onDataChangedCalls;
+    });
```

Comment:
I suggest using a `sinon.spy()` for this

Reply:
Appui doesn't use `sinon`. Changed to use `vi.fn()`.